src/handler/Handler.js

```
@@ -47,47 +52,78 @@ class Handler {
     this.directory = directory;
     this.dependencies = dependencies;
 
-    const categories = fs.readdirSync(directory);
-    categories.forEach((category) => {
-      Utils
-        .readdirSyncRecursive(path.join(directory, category))
-        .filter(file => file.endsWith('.js'))
-        .forEach((file) => {
-          // eslint-disable-next-line global-require, import/no-dynamic-require
-          const CommandOrEvent = require(file);
-
-          if (CommandOrEvent.prototype instanceof Command) {
-            const command = new CommandOrEvent(dependencies);
-            command.category = category;
-
-            this.loadCommand(command);
-          } else if (CommandOrEvent.prototype instanceof Event) {
-            const event = new CommandOrEvent(dependencies);
-            event.category = category;
-
-            this.loadEvent(event);
-          }
-        });
+    // Find and require all JavaScript files
+    const nodes = Utils
+      .readdirSyncRecursive(directory)
+      .filter(file => file.endsWith('.js'))
+      .map(require);
+
+    // Load all Features
+    nodes.forEach((Node) => {
+      if (Node.prototype instanceof Feature) {
+        this.loadFeature(new Node(dependencies));
+      }
+    });
+
+    // Load all Command and Event classes that haven't loaded yet
+    nodes.forEach((Node) => {
+      if (Node.prototype instanceof Command) {
+        const loaded = Array.from(this.commands.values())
+          .some(command => command instanceof Node);
+
+        if (!loaded) {
+          this.loadCommand(new Node(dependencies));
+        }
+      }
+
+      if (Node.prototype instanceof Event) {
+        const loaded = Array.from(this.events.values())
+          .some(events => events
+            .some(event => event instanceof Node));
```

Why the double `some`? Can't you just get the array for the event the Node listens to and then use includes?

---

I see part of the reason is that is the difference in dependency handling. You should at least be able to get rid of one `some` call though.

---

I'm unsure if there's a different way of doing it. Currently it checks if any of the events contains an object which is an instance of `Node`, which is why there's `some` twice.

---

What about creating an instance early and using the `event` or `name` property?

---

I think that would be bad, because then a command or event that should only be instantiated once can be instantiated multiple times. It would be much better to assume each command and event class is only instantiated once, with some occational exceptions.

---

What's the issue creating multiple instances of a command or event? In what scenario should the assumption hold that it will only be instantiated once?

---

Instantiating a command/event might log to console. It firing multiple times could possibly cause confusion. Depending on what the command/event does it might also do some heavier work, which could slow down startup if it's done multiple times. Overall I think it's better to not load things you might not need to load.

---

Doing work in a constructor is an anti pattern, so PRs which include that should be denied anyway. As for logging, what use will logging have when a constructor only creates an object?

---

> As for logging, what use will logging have when a constructor only creates an object?

A command/event that requires some sort of configuration might want to log to console if something is incorrect.

What is a reason for not wanting to hold of loading a command/event until you know it hasn't been loaded yet?

---

if configuration is wrong, you most likely want to throw an exception. As for the reason, it'd make the code a bit easier and faster.

---

> if configuration is wrong, you most likely want to throw an exception.

You have a good point here.

> As for the reason, it'd make the code a bit easier and faster.

The only code it affects this section of the `Handler#load` method. I don't see why it would be faster to create something you might not require, possibly doing unnecessary work.

I also think that if we do it like it currently is, it's more predictable. You (almost) always know a feature loads all it's commands and events, and then the remaining commands and events or loaded. When you load things multiple times in some cases it can become a little more unpredictable. Although this might not cause too many issues, it might be worth taking this into consideration.

---

> The only code it affects this section of the `Handler#load` method. I don't see why it would be faster to create something you might not require, possibly doing unnecessary work.
You also iterate over less events, and creating the instances should be pretty cheap.
> I also think that if we do it like it currently is, it's more predictable. You (almost) always know a feature loads all it's commands and events, and then the remaining commands and events or loaded. When you load things multiple times in some cases it can become a little more unpredictable. Although this might not cause too many issues, it might be worth taking this into consideration.
Creating a second instance of a class should not be unpredictable, unless you're telling me you plan to accept buggy code. You can also remove prototypes of loaded commands / events from the nodes array.

```js
let loaded = new Set([...this.commands.values(), ...this.events.values()].flatMap(c => c.prototype))
nodes.filter(node => loaded.has(node)).forEach(...);
```

---

If we load a command or event multiple times, how will it handle manual dependency injection through features? A feature could provide more values, which the handler itself wouldn't be aware of, thus possibly requiring value checking inside the constructor of a command or event